Function annotations (#1187)
powdr-labs/powdr: book/src/pil/expressions.md

- string literals, written in double quotes, e.g. ``"hello"``
- array literals written in square brackets, e.g. ``[1, 2, 3]``
- tuples, having at least two elements, e.g. `(1, "abc")`
- statement blocks (see below)
- match expressions (see below).
- if expressions (see below).

Parentheses are allowed at any point to force precedence.

### Statement Blocks

A ``{``-``}``-delimited block can be used everywhere where an expression is expected.

It has the form ``{ <statement> ; <statement> ; ... ; <expression> }``,
i.e. a sequence of statements followed by an expression and the value of the statement block
is the value of the final expression.

> The statements can either be expressions or let statements: ``let x = ...`` / ``let x;`

Reviewer: ```suggestion
The statements can either be expressions or let statements: ``let x = ...`` / ``let x;``
```
?

Let statements with value can be used everywhere, they just bind an expression to a local variable
and allow to avoid repeating the expression.

Let statements without value create a new witness column and are only allowed inside [``constr``-functions](#constr-and-query-functions).

Similarly, an expression at statement level can be used to create new constraints that are added to the global constraint set
and this can only be done inside a [``constr``-functions](#constr-and-query-functions).

Note that you can always create constraints and return them from a function, even in [pure function](#constr-and-query-functions).

> Example:

Reviewer: Is this an example of the `Note...` par above?

Author: No, let me move it further up.

```rust
{{#include ../../../test_data/pil/book/generic_to_algebraic.pil}}
```

## Constr and Query Functions

Every function in PIL is either a pure, a `constr` or a `query` function. They are denoted by

- `|...| ...`
- `constr |...| ...`
- `query |...| ...`

Inside `constr` functions, it is possible to create new witness columns
and add constraints to the set of constraints (see the [Statement Blocks](#statement-blocks) section for details).

Inside `query` functions, it is possible to evaluate the value of a column on the "current" row
using the `std::prover::eval` function.

Both actions require a certain context to be available, which is not the case for example when
the values of a fixed column are computed.

A `query` function can only be used in the query or hint part of a witness column while `constr` functions
can only be evaluated in the constraint part of a namespace or machine.

You can define and call new `constr` functions inside a `constr` function and you can call and define

> new `query` functions inside `query` functions, but as soon as you enter a pure function, this is not possible any more.

Reviewer: ```suggestion
new `query` functions inside `query` functions, but as soon as you enter a pure function, this is not possible anymore.
```
